// mango-cms/webpack.config.js
const path = require('path');
const webpack = require('webpack');
const nodeExternals = require('webpack-node-externals');
const NodemonPlugin = require('nodemon-webpack-plugin');


// Get the user's project root from the execution context (passed via CLI or fallback to cwd)
process.env.PROJECT_ROOT = process.env.PROJECT_ROOT || process.cwd();
process.env.MANGO_ROOT = process.env.MANGO_ROOT || __dirname;
let moduleLocation = process.env.PROJECT_ROOT

let legacyMode = process.env.PROJECT_ROOT == process.env.MANGO_ROOT
if (legacyMode) {
    moduleLocation = __dirname
    process.env.PROJECT_ROOT = path.join(__dirname, '..')
}
process.env.CONFIG_ROOT = process.env.PROJECT_ROOT + (legacyMode ? '/config' : '/mango')
console.log('env', process.env.PROJECT_ROOT, process.env.MANGO_ROOT, legacyMode, path.resolve(moduleLocation, 'node_modules'))

const userProjectRoot = process.env.PROJECT_ROOT
const mangoRoot = process.env.MANGO_ROOT
const buildPath = legacyMode ? mangoRoot : userProjectRoot

console.log('moduleLocation', moduleLocation, process.env.CONFIG_ROOT)

module.exports = {
    mode: 'production',
    resolve: {
        alias: {

            '@mango$': path.resolve(mangoRoot, 'src/cms/exports'),
            '@mango': path.resolve(userProjectRoot, legacyMode ? 'config' : 'mango'),

            '@cms': path.resolve(mangoRoot, 'src/cms'),
            '@config': path.resolve(userProjectRoot, legacyMode ? 'config' : 'mango'),
            '@settings': path.resolve(userProjectRoot, legacyMode ? 'config/config/settings.json' : 'mango/config/settings.json'),
            '@fields': path.resolve(mangoRoot, 'src/cms/1. build/fields'),
            '@mongo': path.resolve(mangoRoot, 'src/cms/1. build/libraries/mongo'),
            '@query': path.resolve(mangoRoot, 'src/cms/1. build/libraries/mongo'),
            '@helpers': path.resolve(mangoRoot, 'src/cms/1. build/helpers'),
            '@main': path.resolve(mangoRoot, 'src/cms/2. process/0. main'),
        },
    },
    devtool: 'inline-source-map',
    entry: {
        server: path.resolve(mangoRoot, 'src/main.js'), // Entry point in @mango-cms/core
    },
    output: {
        path: path.resolve(buildPath, 'build'), // Change to build in mango directory
        filename: 'index.js',
    },
    target: 'node',
    plugins: [new NodemonPlugin()],
    node: {
        __dirname: false,
        __filename: false,
    },
    externals: [
        nodeExternals(),
        // Also check for externals in the mango package
        nodeExternals({
            modulesDir: path.resolve(moduleLocation, 'node_modules')
        })
    ],
};
